src/v/redpanda/application.cc

```diff
@@ -716,7 +716,14 @@ void application::wire_up_redpanda_services() {
                 [&c](cloud_storage::configuration cfg) { c = std::move(cfg); });
           })
           .get();
-        construct_service(cloud_storage_api, std::ref(cloud_configs)).get();
+        // 'cloud_storage_api' should be stopped before any object that depends
+        // on it in order to guarantee a graceful shutdown. Because of that a
+        // multi-step initialization has to be used. We need to start the
+        // initialization of the 'cloud_storage_api' first. Then we need to
+        // fully initialize objects that depend on it. Then we need to finish
+        // initialization of the 'cloud_storage_api'
+        start_constructing_service(cloud_storage_api, std::ref(cloud_configs))
+          .get();
```

Comment:
I have one question and related suggestion, if there is an exception thrown between here and `finish_constructing_service` then is cloud_storage_api not added to deferred list? 

Maybe there can be a scope based object so that when it is created it starts the construction process and when destroyed finishes the construction process, to guarantee that cloud_storage_api is always added to deferred list?

like
```
{
  manage_construction_order m(cloud_storage_api, ...); // starts the construction
  ... construct partition recovery manager etc ...
} // destructor of m finishes construction
```

Reply:
if the exception is thrown redpanda won't start

Reply:
@abhijat  IIUC if Redpanda doesn't start, then no actions from _deferred will be called, so it doesn't matter if cloud_storage_api is not in deferred list